lib/sdf-server/src/server/service/action/cancel.rs

```diff
@@ -0,0 +1,37 @@
+use axum::extract::Query;
+use dal::action::{Action, ActionState};
+use dal::{ActionId, Visibility};
+use serde::{Deserialize, Serialize};
+
+use super::ActionResult;
+use crate::server::extract::{AccessBuilder, HandlerContext};
+use crate::service::action::ActionError;
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct PutOnHoldRequest {
+    id: ActionId,
+    #[serde(flatten)]
+    pub visibility: Visibility,
+}
+
+pub async fn cancel(
+    HandlerContext(builder): HandlerContext,
+    AccessBuilder(request_ctx): AccessBuilder,
+    Query(request): Query<PutOnHoldRequest>,
+) -> ActionResult<()> {
+    let ctx = builder.build(request_ctx.build(request.visibility)).await?;
+
+    let action = Action::get_by_id(&ctx, request.id).await?;
+
+    match action.state() {
+        ActionState::Running | ActionState::Dispatched => {
+            return Err(ActionError::InvalidActionCancellation(request.id))
+        }
+        ActionState::Failed | ActionState::OnHold | ActionState::Queued => {}
+    }
```

Comment:
Would like @jhelwig's opinion on this one, i think this makes sense

Reply:
Asking about only being able to cancel Actions that are Failed, On Hold, or Queued? (Cannot do it for Dispatched, or Running.) Yeah, that's correct.

Reply:
Yup, thanks!